### PR TITLE
Implement custom lists in daemon+CLI

### PR DESCRIPTION
**Pull request by Jontified**
Implements custom lists for the daemon and the CLI. Makes sure that the desktop GUI and android gui do not break.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)

### CODE REVIEW

**Review by dlon — CHANGES_REQUESTED** [review]
> Reviewed 21 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: 21 of 23 files reviewed, 6 unresolved discussions (waiting on @Jontified)

___
*[`mullvad-cli/src/cmds/custom_lists.rs` line 44 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzalBL69rOu0KfQZOf:-NXzalBL69rOu0KfQZOg:bqosrai) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-cli/src/cmds/custom_lists.rs#L44)):*
> ```XML
>     Select { name: String },
> 
>     /// Use a random relay from the custom list
> ```

Wrong description.
___
*[`mullvad-daemon/src/lib.rs` line 2455 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzVr_69rdupmU3VPd3:-NXzVr_69rdupmU3VPd4:bm8lj4r) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-daemon/src/lib.rs#L2455)):*
> ```XML
>     };
> 
>     //let mut relay_settings = settings.get_relay_settings();
> ```

Dead code?
___
*[`mullvad-relay-selector/src/matcher.rs` line 52 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzS4nrD18H_EfwKb3K:-NXzS4nrD18H_EfwKb3L:bm8lj4r) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-relay-selector/src/matcher.rs#L52)):*
> ```XML
> 
>     ///// Will replace location from `constraints` with `custom_list_locations`
>     //pub fn new_with_custom_list_locations(
> ```

Dead code?
___
*[`mullvad-types/src/custom_list.rs` line 49 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzJctX1XpOyJWbrQ-u:-NXzJctX1XpOyJWbrQ-v:b-7poy0n) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-types/src/custom_list.rs#L49)):*
> ```XML
>     pub fn new(name: String) -> Self {
>         CustomList {
>             id: Id(uuid::Uuid::new_v4().to_string()),
> ```

Nit: Can we keep this as a `Uuid`? It seems to implement `Serialize`, but maybe there's a `serde` feature.
___
*[`mullvad-types/src/relay_constraints.rs` line 341 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzg763E1h73OeDctEz:-NXzg763E1h73OeDctF-:b-c0s1kd) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-types/src/relay_constraints.rs#L341)):*
> ```XML
>         match self {
>             Self::Location { location } => write!(f, "location: {location}"),
>             // FIXME: This seems awkward, the user shouldn't see the id and instead they probably
> ```

Do users ever see this? Maybe with `mullvad relay get`?
___
*[`mullvad-types/src/settings/mod.rs` line 77 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzUOv21jbXhtt_67ou:-NXzUOv21jbXhtt_67ov:bjcs0oc) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-types/src/settings/mod.rs#L77)):*
> ```XML
>     pub bridge_state: BridgeState,
>     /// All of the custom relay lists
>     #[serde(default)]
> ```

Nit: Is this attribute required? Given that it's already set on the struct.

**Review by dlon — CHANGES_REQUESTED** [review]
> Reviewed 2 of 23 files at r1.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: all files reviewed, 9 unresolved discussions (waiting on @Jontified)

___
*[`mullvad-management-interface/src/types/conversions/custom_list.rs` line 65 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NY25Ma3Ef5_z82XQcaR:-NY25Ma3Ef5_z82XQcaS:bje5vdh) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-management-interface/src/types/conversions/custom_list.rs#L65)):*
> ```XML
>                 };
>                 Self {
>                     state: 0,
> ```

You should be able to cast `State::Add` to `i32` instead.
___
*[`mullvad-management-interface/src/types/conversions/custom_list.rs` line 76 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NY25Vxz4ck7l9fsAg3y:-NY25Vxz4ck7l9fsAg3z:be2u660) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-management-interface/src/types/conversions/custom_list.rs#L76)):*
> ```XML
>                 };
>                 Self {
>                     state: 1,
> ```

You should be able to cast `State::Remove` to `i32` instead.
___
*[`mullvad-management-interface/src/types/conversions/custom_list.rs` line 98 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NY25rIq7CKRm9jEcDtg:-NY25rIq7CKRm9jEcDth:b-bonwcs) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-management-interface/src/types/conversions/custom_list.rs#L98)):*
> ```XML
>                     .ok_or(FromProtobufTypeError::InvalidArgument("missing location"))?,
>             );
>         match custom_list.state {
> ```

You should be able to use `custom_list_location_update::State::from_i32(custom_list.state)`

**Review by Jontified — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: 14 of 23 files reviewed, 3 unresolved discussions (waiting on @dlon)

___
*[`mullvad-daemon/src/lib.rs` line 2455 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzVr_69rdupmU3VPd3:-NY2DXZqCfkqZpeJ6GuX:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-daemon/src/lib.rs#L2455)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Dead code?
</blockquote></details>

Done.
___
*[`mullvad-management-interface/proto/management_interface.proto` line 578 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzCq9W9IntZynLH-W3:-NY2DZGFDgJeiOaJ8zz1:bf3n68a) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-management-interface/proto/management_interface.proto#L578)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

We'll need an event for custom lists, right? Since the CLI needs to be able to notify the GUI of any changes to them.

But as long as the GUI doesn't support them, I guess that this event could be added in a separate PR.
</blockquote></details>

Every action that changes a custom list will notify about a settings change. So this shouldn't be necessary.
___
*[`mullvad-management-interface/src/types/conversions/custom_list.rs` line 98 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NY25rIq7CKRm9jEcDtg:-NY2G51iAxBGExo2t518:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-management-interface/src/types/conversions/custom_list.rs#L98)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

You should be able to use `custom_list_location_update::State::from_i32(custom_list.state)`
</blockquote></details>

Done.

**Review by dlon — CHANGES_REQUESTED** [review]
> Reviewed 9 of 9 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: all files reviewed, 1 unresolved discussion (waiting on @Jontified)

___
*[`mullvad-types/src/relay_constraints.rs` line 341 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzg763E1h73OeDctEz:-NY3GkVMB48Vh9dYd1Mz:b-gehlvt) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-types/src/relay_constraints.rs#L341)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Do users ever see this? Maybe with `mullvad relay get`?
</blockquote></details>

Since you see it with `mullvad relay get`, we should probably make that CLI command not use `Display` to print the constraints.

**Review by Jontified — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: 16 of 23 files reviewed, 1 unresolved discussion (waiting on @dlon)

___
*[`mullvad-types/src/relay_constraints.rs` line 341 at r1](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NXzg763E1h73OeDctEz:-NYO7f7_0X7wukkJW4NB:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/a5a5f9c35f7f41bbc4e29ac1f2ead2e2897126b0/mullvad-types/src/relay_constraints.rs#L341)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Since you see it with `mullvad relay get`, we should probably make that CLI command not use `Display` to print the constraints.
</blockquote></details>

Done.

File: gui/src/main/daemon-rpc.ts [hunk 1]
```diff
@@ -1233,15 +1233,25 @@ function convertFromConnectionConfig(
...
-  }
-  if (location.city) {
-    return { city: [location.country, location.city] };
+function convertFromLocation(location: grpcTypes.LocationConstraint.AsObject): RelayLocation {
+  // FIXME: This is a hack that assumes that the LocationConstraint is not a custom list.
+  // If it is we just set the country to "any" even if that isn't correct.
+  if (location.location == undefined) {
+    return { country: "any" };
   }
+  else {
```
Comment (MrChocolatine) [inline gui/src/main/daemon-rpc.ts:1242]: `else` branch is extraneous.

```suggestion
```

**Review by dlon — CHANGES_REQUESTED** [review]
> Reviewed 10 of 11 files at r3, 4 of 4 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: all files reviewed, 2 unresolved discussions (waiting on @dlon and @Jontified)

___
*[`mullvad-types/src/relay_constraints.rs` line 416 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NYrGqIqE5ZldkrCPH_t:-NYrGqIqE5ZldkrCPH_u:bi4353s) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2264685d8ab94e788ca15ad2c7c3ca51624d1f92/mullvad-types/src/relay_constraints.rs#L416)):*
> ```XML
>                 writeln!(
>                     f,
>                     "Tunnel protocol: Any\nOpenVPN: {}\nWireguard: ",
> ```

For consistency, replace `OpenVPN` and `Wireguard` with `OpenVPN constraints` and `Wireguard constraints`.

**Review by dlon — APPROVED** [review]
> Reviewed 10 of 12 files at r5, 10 of 10 files at r6, all commit messages.
> Dismissed @MrChocolatine from a discussion.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: :shipit: complete! all files reviewed, all discussions resolved

**Review by Jontified — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: 20 of 26 files reviewed, all discussions resolved (waiting on @dlon)

___
*[`gui/src/main/daemon-rpc.ts` line 1242 at r3](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#gh-1241141355:-NYrQdHFC8JQgWjaVnEi:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/d66909569be2ed690ac146bf7e07d976ca4bcd57/gui/src/main/daemon-rpc.ts#L1242)):*
<details><summary><i>Previously, MrChocolatine wrote…</i></summary><blockquote>

`else` branch is extraneous.

```suggestion
```
</blockquote></details>

Done.
___
*[`mullvad-types/src/relay_constraints.rs` line 416 at r4](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NYrGqIqE5ZldkrCPH_t:-NYrQuvw3SL8gbllyOj-:b-896fix) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/2264685d8ab94e788ca15ad2c7c3ca51624d1f92/mullvad-types/src/relay_constraints.rs#L416)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

For consistency, replace `OpenVPN` and `Wireguard` with `OpenVPN constraints` and `Wireguard constraints`.
</blockquote></details>

Done.

**Review by dlon — CHANGES_REQUESTED** [review]
> Reviewed 3 of 9 files at r7, 2 of 9 files at r8, 1 of 3 files at r9, 8 of 15 files at r10, 4 of 4 files at r11, 8 of 8 files at r12, 1 of 1 files at r13, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: all files reviewed, 1 unresolved discussion (waiting on @Jontified)

___
*[`mullvad-daemon/src/settings.rs` line 378 at r13](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NZ17DFt85CqfGSnhcg5:-NZ17DFt85CqfGSnhcg6:bfpjbii) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e804caf1eec0e5ec32bbd75a0abd146ff755176e/mullvad-daemon/src/settings.rs#L378)):*
> ```XML
>               "relay_settings": {
>                 "normal": {
>                   "location": {
> ```

Settings migration code is needed if we can't get the serialization to be identical.

**Review by Jontified — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: 15 of 43 files reviewed, 1 unresolved discussion (waiting on @dlon)

___
*[`mullvad-daemon/src/settings.rs` line 378 at r13](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NZ17DFt85CqfGSnhcg5:-NZ1lt_F1Ss8MLeW3gnR:bmwr5vb) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/e804caf1eec0e5ec32bbd75a0abd146ff755176e/mullvad-daemon/src/settings.rs#L378)):*
<details><summary><i>Previously, dlon (David Lönnhager) wrote…</i></summary><blockquote>

Settings migration code is needed if we can't get the serialization to be identical.
</blockquote></details>

Added settings migration code

**Review by dlon — APPROVED** [review]
> Reviewed 3 of 24 files at r14, 6 of 15 files at r15, 2 of 4 files at r16, 7 of 8 files at r17, 9 of 9 files at r18, 1 of 1 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: all files reviewed, 1 unresolved discussion (waiting on @Jontified)

___
*[`mullvad-daemon/src/migrations/v6.rs` line 83 at r19](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797#-NZ5A9Fx0E9KLAKjwLvC:-NZ5A9Fx0E9KLAKjwLvD:b-vdtbw9) ([raw file](https://github.com/mullvad/mullvadvpn-app/blob/9c6fc569b9f1846b9565f464bbf8ac52252aa494/mullvad-daemon/src/migrations/v6.rs#L83)):*
> ```XML
>         .and_then(|normal_relay_settings| normal_relay_settings.get_mut("location"))
>     {
>         if let Some(only) = location.get_mut("only") {
> ```

Nit: This body could be moved to a function and used in all three places.

**Review by dlon — APPROVED** [review]
> Reviewed 1 of 1 files at r20, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/mullvad/mullvadvpn-app/4797)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**